18 add version control guidance language agnostic mw comments and tidy up (#53)
nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook: guidance/02-develop/pages/03-version-controlled.qmd

description: "Version control is all about managing changes"
---

## At a glance
Version control (Git) helps teams to safely collaborate and make their code resilient by tracking changes, enabling branching, and ensuring code is transparent, auditable and reproducible. 

This guidance aims to give users an understanding of version control, the benefits of its use, and how to implement it in their work.

## What is version control?

Version control is a way to track line-by-line changes to source code over time to give a complete history of edits to a file. It supports collaboration on analytical projects by managing different versions of the same files. It also allows any point in history to be restored. 

Version control helps us to solve the problem of managing file versions manually through inefficient naming conventions and [keep an audit trail of changes](https://best-practice-and-impact.github.io/qa-of-code-guidance/version_control.html#keep-an-audit-trail-of-changes).

> Version control allows us to track changes in source code and its accompanying documentation and dependencies, streamlining environment and package management.

Reviewer: **(Optional, wordsmithing):** just a suggestion for flow - roughly the same words but different order

```suggestion
Version control is a way to track line-by-line changes to source code over time. Version control [keeps an audit trail of changes](https://best-practice-and-impact.github.io/qa-of-code-guidance/version_control.html#keep-an-audit-trail-of-changes) and allows any point in history to be restored. 

Version control supports collaboration on analytical projects by managing different versions of the same files and helps us to solve the problem of managing file versions manually through inefficient naming conventions. 

As well as tracking the history of source code, version control helps us track changes in accompanying documentation and dependencies, streamlining environment and package management.
```

There are 2 main ways to interact with git: 

* through the command line interface (Git Bash) - the Git CLI is more difficult to use for beginners but much more flexible and powerful 

* through IDEs with in-built Git functionality - tools like R Studio and Visual Studio Code help users get started with the basics of creating repositories, linking them to a remote, committing, pushing, and pulling through streamlined GUIs. 

Our [suggested learning resources](#learn-more) provide additional support for beginner and expert Git users.

## Why should we use version control?
Version control is a key tool for assuring analytical code. It enables teams to manage change safely, track code lineage, collaborate, and reproduce results with confidence.

Version control makes code:

* transparent and auditable by keeping a complete, attributed history of every change for full traceability
* reproducible, making sure any past version of the code can be restored

> * safe to collaborate and work in parallel on through branching, preventing clashes and enforcing peer review

Reviewer: **(Suggestion, blocking):** add a link to peer review

* easy to revert to a stable version after problematic changes
* resilient by storing work safely in a remote repository, protecting against local failures.
* easier to share knowledge about, providing clear history and structure to help new team members understand the codebase

## How do we use version control?

Teams should adopt tools and processes that fit their circumstances. However, when you adopt version control we recommend following a set approach:

* commit often -- small, frequent commits with meaningful messages are easier to review and understand for both other team members and your future self

* push to the remote regularly -- committing to your local repository doesn't mean that you changes are reflected for everyone else. Remember to push your commits to the remote

* use branching -- use a branching strategy to organise work and separate changes by collaborators. Your `main` branch should always be assured, reviewed code that can be trusted

> * peer review -- use pull/merge request functionality from Git hosting providers to review code before merging changes

Reviewer: **(Suggestion, blocking):** as above - link to peer review

* manage sensitive data -- follow best practice when managing credentials, secrets, API keys, and personal data. It is easy to accidentally check these things in to version control and difficult to remove them again

:::{.callout-tip}
## Rule of thumb
__Commit__ after each small task is completed, such as a function, writing a common table expression, or developing a visualisation. __Push__ to the remote at least once a day.
:::

Git has a learning curve and it is easy to make mistakes when getting started but don't be afraid to break things. Git is designed to make fixing things easy and mistakes are usually reversible. 

### Setting Git up
Git uses a configuration file to set your name and email address that are attached to commits. It is essential that you set your Git config to allow changes and commits to be attributable to you as an author. 

Guidance on setting your config and how user accounts for git providers should be created are available on the [NHSBSA Confluence](https://bsa2468.atlassian.net/wiki/spaces/KB/pages/5801672736/Github+User+Accounts).

### Commits

> A commit records the changes made to the file(s) that you're working on. The message attached to a commit should clearly communicate what changed and why. Most commit messages are short and informative and should be written in the imperative. For example, 'Add feature' or 'Fix bug', not 'Added feature', 'Adds feature', or 'Fixed bug', 'Fixes bug':

Reviewer: **(Suggestion, non-blocking)**: just a small suggestion for clarity 

```suggestion
A commit records the changes made to the file(s) that you're working on. The message attached to a commit should clearly communicate what changed and why. Most commit messages are short and informative and should be written in the imperative, i.e. how the commit changes the codebase in the present tense, rather than what you did in the past tense. For example, 'Add feature', not 'Added feature' or 'Adds feature', and 'Fix bug' not 'Fixed bug' or 'Fixes bug':
```

```bash

> add table creation function

Reviewer: **(Suggestion, blocking):** capitalised for consistency with guidance above

```suggestion
Add table creation function
```

```

```bash

> fix bug in identified_patient flag

Reviewer: **(Suggestion, blocking):** capitalised for consistency with guidance above

```suggestion
Fix bug in identified_patient flag
```

```

In some cases you may want to provide more detail, the first line of a commit message can be treated like the subject of an email with the rest of the text as the body.

You should commit frequently when you have completed a small, discrete amount of work. For example, you have developed a chart, created a table, or written an output. 

Small, frequent commits make understanding the history of a file much easier and reduces the burden for a peer reviewer. Small commits also make it easier to roll back changes after the introduction of a bug or issue -- large, monolithic commits make bug hunting and fixing harder. 

### Pushing

> Changes that you make to your local version of the files (repository) remain local until you explicitly "push" them to the remote. a Remote is typically a repository hosted by one of the big Git providers -- GitHub, GitLab, or Azure DevOps -- that has been synced to your local copy. If you "clone" a repository from a service this syncing is done automatically. If you have created a local repository first you will have to set the remote manually.

Reviewer: **(Typo, blocking):** 

```suggestion
Changes that you make to your local version of the files (repository) remain local until you explicitly "push" them to the remote. A remote is typically a repository hosted by one of the big Git providers -- GitHub, GitLab, or Azure DevOps -- that has been synced to your local copy. If you "clone" a repository from a service this syncing is done automatically. If you have created a local repository first you will have to set the remote manually.
```

> Pushing your changes to the remote makes them available to everyone to "pull" into their local copies. Pushing should be done regularly, typically at least once a day, or more frequently if you are making rapid, large scale changes that others need to progress their work.

Reviewer: **(Suggestion, non-blocking):** could add something about generally not pushing to main?

Reviewer: Or to the bit about branching below

git push origin v1.3.0
```

Once a tag has been created, these locations in a project's history can be easily found and recovered.

### Commits

Earlier in this guidance we defined the simple approach to commits and writing commit messages. As your project grows you may find that your Git history growing with it and becoming more complex, with the previous approach to commits no longer being fit for purpose.

#### Logical commits

Logical commits are well defined units of work that represent a single coherent change to the codebase. They:

* address a single concern or feature
* can be understood on their own without requiring context from other commits
* are complete enough to not break functionality
* clearly explain what changed and why using a commit message

For example, a logical commit could be the full unit of work required to develop a data visualisation or a data aggregation function. 

Take care not to sacrifice logical commits by committing too frequently. If you wish to commit unfinished work, then make this clear by labelling the commit as work in progress (wip) and later [squash](#squashing) the commits.

#### Squashing

> We recommend squashing commits into logical commits prior to merging downstream to branches such as `main` or `develop`. This removes unneccesary commits that can confuse, providing a clean and logical git history. Use [interactive rebase][atlassian-git-rebase] to take control of squashing, and the resulting commit messages.

Reviewer: **(Question, blocking)** Not sure about interactive rebase as advice here? Are we saying squash or rebase or both?

Author: I agree, was adapting guidance for developers, but think this isn't necessary and confuses other guidance. will add an issue

Most Git hosting providers support squashing on merge. We recommend against this. Automating removes control from authors on the resulting commit message. There is a risk that detail is lost from the squashed commits, or too much needless information is included. If you are using an issue tracker and adding issue IDs to your commit messages, there is a further risk that issue IDs are lost, removing audit.

> #### Conventional Commits

Reviewer: **(Suggestion, blocking):** is this too much detail? Too prescriptive? Do we think in terms of fix/feature? 

Also minor point but think the verbs should be capitalised (e.g. `Fix` not `fix`) to align with other guidance above.

Author: covered as part of https://github.com/nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook/issues/60

* `docs` documentation only changes
* `feat` a new feature
* `fix` a bug fix
* `perf` a code change that improves performance
* `refactor` a code change that neither fixes a bug nor adds a feature
* `style` changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc). We always recommend separating formatting changes from functional changes
* `test` adding missing tests or correcting existing tests
* `revert` backing out a previous change

You can denote a breaking change in your commit by adding `!` after the type or scope in your commit message. Breaking changes can be part of commits of any type.
</details>

A typical conventional commit looks like:

```html
<type>[optional scope]: <description>

[optional body]

[optional footer(s)]
```

The commit header should contain the type of change, ticket reference if using an issue tracker, and a short description. For example:

```bash
feat: OST-123 adds time series chart
```

The optional body should contain details of specific changes if required. The optional footer should contain [git-trailer][git-trailer-info] formatted meta-information

> ### Rebasing

Reviewer: **(Suggestion, blocking):** again think we need to discuss - not sure we want to recommend rebasing. 

Could we not just recommend that people use branches that represent discrete pieces of work and describe the changes in a sufficient level of detail via PR? Rather than rebase which adds complexity for less experienced git users

Author: covered as part of https://github.com/nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook/issues/60
